### PR TITLE
fixing obsolete EDAnalyzer in DQM/BeamMonitor

### PR DESCRIPTION
**Pull request by emanueleusai**
#### PR description:

replacing EDAnalyzer with the appropriate edm::one version for various plugins in the DQM/BeamMonitor package.
No changes to ME expected.

#### PR validation:

DQM/BeamMonitor now compiles on CMSSW_12_3_CMSDEPRECATED_X_2022-02-21-2300 without warnings

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport

### CODE REVIEW

**Comment by emanueleusai** [comment]
assign alca

**Comment by emanueleusai** [comment]
addresses part of #36404

**Comment by mmusich** [comment]
There's an earlier attempt from @francescobrivio at #35501. We've been wondering if a `DQMEDAnalyzer` type should not be used instead: https://github.com/cms-sw/cmssw/pull/35501#issuecomment-932213550
Perhaps you can clarify.

**Comment by emanueleusai** [comment]
@mmusich Interesting. Somehow I forgot about @francescobrivio 's previous attempt. Anyway the deadline for the transition is now imminent so I think it's time to discuss this again.

I implemented it in a slightly different way wrt Francesco, so let's see the result of the tests.

Concerning the use of DQMOneEDAnalyzer, I largely agree with the discussion in the previous PR. Perhaps one can create a new DQMOne"LumiAndRun"EDAnalyzer that templates both edm::one::WatchLuminosityBlocks and edm::one::WatchRuns. But then one needs to rewrite AlcaBeamMonitorClient & others to access DQMStore through the DQMOne*EDAnalyzer interface rather than accessing it directly. I think just changing the class type isn't going to work.

**Comment by emanueleusai** [comment]
please test

File: DQM/BeamMonitor/plugins/BeamConditionsMonitor.h [hunk 1]
```diff
@@ -36,7 +36,7 @@ class BeamConditionsMonitor : public edm::EDAnalyzer {
   void beginJob() override;
 
   // BeginRun
-  void beginRun(const edm::Run& r, const edm::EventSetup& c) override;
+  // void beginRun(const edm::Run& r, const edm::EventSetup& c) override;
```
Comment (tvami) [inline DQM/BeamMonitor/plugins/BeamConditionsMonitor.h:39]: Should these be removed for good? (also the `endRun`)

Comment (emanueleusai) [reply DQM/BeamMonitor/plugins/BeamConditionsMonitor.h:39]: yup I forgot to remove them after testing

**Comment by emanueleusai** [comment]
please abort test

**Comment by emanueleusai** [comment]
please test

**Comment by mmusich** [comment]
> I implemented it in a slightly different way wrt Francesco, so let's see the result of the tests.

Let me rephrase. What's the preferred strategy from the DQM group? Clearly the plain old `EDAnalyzer` class that we have now is still technically working, but there must be some advantage computing-wise in using the producer-based types. Aren't `one` modules notouriously bad at threading efficiency?

> others to access DQMStore through the DQMOne*EDAnalyzer interface rather than accessing it directly. I think just changing the class type isn't going to work.

Right, I admit I didn't follow much further Francesco's PR, but clearly just changing the method signatures is not gonna cut it, the `iBooker` and `iGetter` objects shall be used for interacting with the `DQMStore`

File: DQM/BeamMonitor/plugins/AlcaBeamMonitorClient.cc [hunk 2]
```diff
@@ -70,6 +70,8 @@ void AlcaBeamMonitorClient::beginRun(const edm::Run& r, const EventSetup& contex
 void AlcaBeamMonitorClient::analyze(const Event& iEvent, const EventSetup& iSetup) {}
 
 //----------------------------------------------------------------------------------------------------------------------
+void AlcaBeamMonitorClient::beginLuminosityBlock(const LuminosityBlock& iLumi, const EventSetup& iSetup) {}
```
Comment (francescobrivio) [inline DQM/BeamMonitor/plugins/AlcaBeamMonitorClient.cc:73]: Not related to this line, but shouldn't the `~AlcaBeamMonitorClient() {}` be changed to  `~AlcaBeamMonitorClient() = default;` ?

Comment (francescobrivio) [reply DQM/BeamMonitor/plugins/AlcaBeamMonitorClient.cc:73]: the same comment can be applied to 
- BeamConditionsMonitor.cc
- PixelVTXMonitor.cc
- Vx3DHLTAnalyzer.cc

**Comment by mmusich** [comment]
there's plenty of failures, presumably coming from the inclusion of the commits of https://github.com/cms-sw/cmssw/pull/37034 in this PR.

**Review by francescobrivio — COMMENTED** [review]
I agree with @emanueleusai that a new DQMStore class could be used, but that would require a larger work to update all the plugins, maybe it can be posponed to later given that technically the `one:EDAnalyzer` class is technically ok for the moment.
As @mmusich said @cms-sw/dqm-l2 maybe you can comment on what is the preferred solution from DQM point of view?

**Comment by emanueleusai** [comment]
@mmusich @francescobrivio We discussed this internally with the other @cms-sw/dqm-l2 . Our preferred position would be to first move from EDAnalyzer to one::EDAnalyzer or similar (e.g. DQMEDAnalyzer to DQMOneEDAnalyzer) in order to comply with the deprecation of EDAnalyzer. Then later on, in a separate set of PRs, address the issue of the DQM modules still not using DQM*EDAnalyzer and make sure DQMStore is not accessed directly, but only through the iBooker/iGetter methods.

File: DQM/BeamMonitor/plugins/AlcaBeamMonitorClient.cc (hunk 2 above)
Comment (emanueleusai) [reply DQM/BeamMonitor/plugins/AlcaBeamMonitorClient.cc:73]: good point, I'll fix that too

**Comment by mmusich** [comment]
@cmsbuild, please test
   * to get a clean comparison

**Comment by makortel** [comment]
From `core` point of view it would be preferable to address the real issues at the same time the base class type is changed. On a quick peek on DQMStore and DQMEDAnalyzer base classes I was not able to convince myself if it would be safe to use DQMStore directly from two non-DQMEDAnalyzer modules concurrently.

**Comment by emanueleusai** [comment]
@makortel thank you for giving the core point of view on this. I can make an attempt at rewriting the module(s) to use DQM*EDAnalyzer but it'd be safer if BeamMonitor experts (AlCa developers I assume) can look into it if we go this way.

**Comment by francescobrivio** [comment]
> @makortel thank you for giving the core point of view on this. I can make an attempt at rewriting the module(s) to use DQM*EDAnalyzer but it'd be safer if BeamMonitor experts (AlCa developers I assume) can look into it if we go this way.

I am the only developer at the moment and I would need some time to (digest and) test the correct way to update these plugins (maybe following Javi's suggestion from https://github.com/cms-sw/cmssw/pull/35501#issuecomment-942007011 since my last attempt with DQM*EDAnalyzer wasn't successful).
IIUC the current implementation from @emanueleusai is technically correct although not the preferred one from `core`.
So I guess the question is back to @cms-sw/dqm-l2 and @makortel on they urgency of this update? 
The options are:
- we merge this one moving away from deprecated APIs now and posponing the update to DQM*EDAnalyzer to later
- we reject this one keep having the deprecated EDAnalyzer till I have time to fix it all in one go

**Comment by mmusich** [comment]
> I am the only developer at the moment 

I guess I can help with that.

> we merge this one moving away from deprecated APIs now and posponing the update to DQM*EDAnalyzer to later

IMHO this solution is equivalent to sweeping the dust under the rug. The deprecation warning is gone, but then there's no guarantee it won't fail in production.

**Comment by francescobrivio** [comment]
-1

 - After private conversation with @mmusich we decided to give it a try and do all the changes at the same time as soon as possible. So we can close this.
 - @makortel could you remind me the timelines for moving away from deprecated EDAnalyzer?

**Comment by makortel** [comment]
> * After private conversation with @mmusich we decided to give it a try and do all the changes at the same time as soon as possible. So we can close this.

Thanks!

> * @makortel could you remind me the timelines for moving away from deprecated EDAnalyzer?

The goal to migrate (at minimum) all legacy EDModules being used in production is by 13_0_0 (at which point we'll change the current warning message of such modules to an exception).
